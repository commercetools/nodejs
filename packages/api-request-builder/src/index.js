// eslint-disable-next-line import/prefer-default-export
export { default as createRequestBuilder } from './create-request-builder'
export * as features from './features'
