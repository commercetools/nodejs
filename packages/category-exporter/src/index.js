import CategoryExporter from './main'

export default CategoryExporter
