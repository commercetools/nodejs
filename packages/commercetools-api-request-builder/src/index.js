export { default } from './create-request-builder'
export * as features from './features'
