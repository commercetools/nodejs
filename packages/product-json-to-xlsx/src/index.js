export default from './main'
export { writeToSingleXlsxFile, writeToZipFile } from './writer'
export { default as ProductExcel } from './product-excel'
