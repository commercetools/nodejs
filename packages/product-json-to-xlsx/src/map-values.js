/* @flow */

type HeaderObj = {
  label: string,
  value: string,
}

export default function mapValues(
  headers: Array<HeaderObj>,
  mappedProduct: Object
): Array<any> {
  return headers.map(
    (header: HeaderObj): any => mappedProduct[header.value] || ''
  )
}
