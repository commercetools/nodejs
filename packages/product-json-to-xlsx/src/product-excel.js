/* @flow */
import Excel from 'exceljs'

interface Stream {}
type WorkbookWriterOptions = {
  stream: Stream,
  filename: string,
  useSharedStrings: boolean,
  useStyles: boolean,
}
type WorkbookWriter = {
  constructor(options: $Shape<WorkbookWriterOptions>): WorkbookWriter,
  // commit all worksheets, then add suplimentary files
  commit(): Promise<void>,
  addStyles(): Promise<void>,
  addThemes(): Promise<void>,
  addOfficeRels(): Promise<void>,
  addContentTypes(): Promise<void>,
  addApp(): Promise<void>,
  addCore(): Promise<void>,
  addSharedStrings(): Promise<void>,
  addWorkbookRels(): Promise<void>,
  addWorkbook(): Promise<void>,
}
type Column = {
  header: string | string[],
  key: string,
  width: number,
  outlineLevel: number,
  hidden: boolean,
}
type Row = {
  height: number,
  commit(): void,
}
type Worksheet = {
  addRow(data: any[] | any): Row,
  columns: Array<$Shape<Column>>,
}
type ExcelConfig = {
  workbook: WorkbookWriter,
  worksheet: Worksheet,
}

export default class ProductExcel {
  // Set flowtype annotations
  stream: stream$Writable
  excel: ExcelConfig

  constructor(output: stream$Writable, worksheetName: string = 'Products') {
    this.stream = output
    this.excel = ProductExcel.init(output, worksheetName)
  }

  static init(
    outputStream: stream$Writable,
    worksheetName: string
  ): ExcelConfig {
    const workbookOpts: Object = {
      stream: outputStream,
      useStyles: true,
      useSharedStrings: true,
    }

    const workbook = new Excel.stream.xlsx.WorkbookWriter(workbookOpts)
    const worksheet = workbook.addWorksheet(worksheetName)
    return { workbook, worksheet }
  }

  /**
   * Replace undefined and empty string with null values
   * @param row Array of values
   */
  static _cleanRow(row: Array<any> | any): Array<any> {
    return row.map(item =>
      typeof item === 'undefined' || item === '' ? null : item
    )
  }

  writeHeader(header: Array<string> | string) {
    // eslint-disable-next-line no-param-reassign
    this.excel.worksheet.columns = Array.isArray(header)
      ? header.map(name => ({ header: name }))
      : [{ header }]
  }

  writeRow(row: Array<any>) {
    this.excel.worksheet.addRow(ProductExcel._cleanRow(row)).commit()
  }

  finish(): Promise<void> {
    return this.excel.workbook.commit()
  }
}
