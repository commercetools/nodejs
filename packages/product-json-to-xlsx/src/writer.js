/* @flow */
import fs from 'fs'
import tmp from 'tmp'
import path from 'path'
import slugify from 'slugify'
import {
  mapHeaders,
  onStreamsFinished,
  archiveDir,
} from '@commercetools/product-json-to-csv'
import type { highlandStream } from 'types/highland'

import mapValues from './map-values'
import ProductExcel from './product-excel'

// Accept a highland stream and write the output to a single file
export function writeToSingleXlsxFile(
  productStream: highlandStream,
  output: stream$Writable,
  logger: Object,
  headers: Array<Object>
) {
  const mappedHeaders = mapHeaders(headers)
  const headerNames = mappedHeaders.map((header) => header.label)

  const excel = new ProductExcel(output)
  excel.writeHeader(headerNames)

  productStream
    .each((product) => {
      const row = mapValues(mappedHeaders, product)
      if (row.join('')) excel.writeRow(row)
    })
    .done(() => {
      // finish workbook and close the stream
      excel.finish()
      logger.info('All products have been written to XLSX file')
    })
}

export function finishWorksheetsAndArchive(
  exports: Array<Object>,
  dir: string,
  output: stream$Writable,
  logger: Object
) {
  if (exports.length === 0) return archiveDir(dir, output, logger)

  const writeStreams: Array<stream$Writable> = exports.map(
    (exportInfo) => exportInfo.excel.stream
  )
  onStreamsFinished(writeStreams, (err) => {
    if (err) {
      logger.error(err)
      output.emit('error', err)
    } else {
      archiveDir(dir, output, logger)
      logger.info('All products have been written to ZIP file')
    }
  })

  // close all excel workbooks
  return exports
    .map((exportInfo: Object): Object => exportInfo.excel)
    .forEach((excel: Object) => excel.finish())
}

// Accept a highland stream and write the output to multiple XLSX files per
// product type, then compress all files to a zip file
export function writeToZipFile(
  productStream: Object,
  output: stream$Writable,
  logger: Object
): void {
  const tmpDir: string = tmp.dirSync({ unsafeCleanup: true }).name
  tmp.setGracefulCleanup()

  const exportByProductType: Object = {}
  let lastExport
  let lastProductType

  productStream
    .each((product: Object): void => {
      // check what productType are we exporting
      if (product.productType && product.productType !== lastProductType) {
        lastProductType = product.productType
        lastExport = exportByProductType[lastProductType]

        // if we haven't started exporting this productType yet
        if (!lastExport) {
          // generate a temp file name
          const fileName: string = `${slugify(product.productType)}.xlsx`
          const filePath: string = path.join(tmpDir, fileName)

          // create headers and file stream
          lastExport = {
            productType: lastProductType,
            headers: mapHeaders(Object.keys(product)),
            excel: new ProductExcel(fs.createWriteStream(filePath)),
          }

          // write a header row to XLSX file
          const headerNames: Array<string> = lastExport.headers.map(
            (header) => header.label
          )
          lastExport.excel.writeHeader(headerNames)

          // register new export in cache
          exportByProductType[lastProductType] = lastExport
        }
      }

      // we have lastExport variable containing all necessary info
      lastExport.excel.writeRow(mapValues(lastExport.headers, product))
    })
    .done((): void => {
      const exports: Array<Object> = Object.keys(exportByProductType).map(
        (key) => exportByProductType[key]
      )
      finishWorksheetsAndArchive(exports, tmpDir, output, logger)
    })
}
