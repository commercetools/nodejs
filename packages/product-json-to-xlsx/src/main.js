/* @flow */
import type { highlandStream } from 'types/highland'
import ProductJsonToCsv from '@commercetools/product-json-to-csv'
import { writeToSingleXlsxFile, writeToZipFile } from './writer'

export default class ProductJsonToXlsx extends ProductJsonToCsv {
  run(input: stream$Readable, output: stream$Writable) {
    const productStream: highlandStream = this.parse(input, output)
    const { headerFields } = this.parserConfig

    if (headerFields)
      writeToSingleXlsxFile(productStream, output, this.logger, headerFields)
    else writeToZipFile(productStream, output, this.logger)
  }
}
