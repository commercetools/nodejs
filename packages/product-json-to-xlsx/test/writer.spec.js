import fs from 'fs'
import highland from 'highland'
import Excel from 'exceljs'
import tmp from 'tmp'
import StreamTest from 'streamtest'
import unzipper from 'unzipper'
import * as writer from '../src/writer'

tmp.setGracefulCleanup()

const streamTest = StreamTest.v2

function analyzeExcelWorkbook(workbook) {
  const rows = []
  const worksheet = workbook.getWorksheet('Products')

  worksheet.eachRow((row) => rows.push(row.values))
  // remove first column containing null values
  rows.forEach((row) => row.shift())

  return {
    workbook,
    worksheet,
    rows,
  }
}

async function analyzeExcelFile(path) {
  const workbook = new Excel.Workbook()
  await workbook.xlsx.readFile(path)
  return analyzeExcelWorkbook(workbook)
}

function analyzeExcelStream(stream) {
  const workbook = new Excel.Workbook()
  const readStream = workbook.xlsx.createInputStream()
  stream.pipe(readStream)

  return new Promise((resolve, reject) => {
    readStream.on('error', reject)
    readStream.on('done', () => resolve(analyzeExcelWorkbook(workbook)))
  })
}

describe('Writer', () => {
  let logger
  let sampleProducts
  beforeEach(() => {
    logger = {
      info: jest.fn(),
      error: jest.fn(),
    }
    sampleProducts = [
      {
        id: '12345ab-id',
        key: 'product-key',
        productType: 'product-type-1',
        'variant.id': 1,
        'variant.sku': 'A0E200000001YKI',
        'variant.images': 'image-url1',
        addedAttr: '',
        anotherAddedAttr: '',
        article: 'sample 089 WHT',
        color: 'white',
        'colorFreeDefinition.en': 'black-white',
      },
      {
        'variant.id': 6,
        'variant.sku': 'A0E200001YKI123',
        'variant.images': 'https://image1.de;myImage.com',
        article: 'sample 089 WHT',
        color: 'white',
        'colorFreeDefinition.en': 'black-white',
      },
      {
        id: '75345ab-id',
        key: 'product-key-2',
        productType: 'product-type-2',
        'variant.id': 1,
        'variant.sku': 'A0E200000001YKI',
        'variant.images': 'https://example.com/foobar/commer.jpg|3|4',
        addedAttr: '',
        anotherAddedAttr: '',
        article: 'sample 089 WHT',
        color: 'white',
        'colorFreeDefinition.en': 'black-white',
        'colorFreeDefinition.de': 'schwarz-weiß',
        designer: 'michaelkors',
      },
      {
        'variant.id': 3,
        'variant.sku': 'A0E200001YKI123',
        'variant.images': 'https://example.com/foobar/commer234.jpg|3|3',
        'colorFreeDefinition.de': 'schwarz-weiß',
        designer: 'michaelkors',
      },
      {
        id: 'another5ab-id',
        key: 'product-key',
        productType: 'product-type-1',
        'variant.id': 1,
        'variant.sku': 'A0E200001YKI',
        addedAttr: '',
        anotherAddedAttr: '',
        article: 'sample 777 WHT',
        color: 'grune',
        'colorFreeDefinition.en': 'grey-white',
      },
      {
        'variant.id': 21,
        'variant.sku': 'PPP0001YKI123',
        'variant.images': 'https://eg993',
        'colorFreeDefinition.en': 'schwarz-weiß',
        color: 'blau',
      },
    ]
  })

  describe('::writeToSingleXlsxFile', () => {
    test('write products to a single file with specified headers', async () => {
      return new Promise((done) => {
        const sampleStream = highland(sampleProducts)
        const tempFile = tmp.fileSync({ postfix: '.xlsx', keep: true })
        const output = tempFile.name
        const outputStream = fs.createWriteStream(output)
        const headers = [
          'id',
          'key',
          'productType',
          'variantId',
          'sku',
          'images',
          'anotherAddedAttr',
          'article',
          'designer',
          'color',
        ]

        outputStream.on('error', done)
        outputStream.on('finish', async () => {
          const { workbook, worksheet, rows } = await analyzeExcelFile(output)
          const secondWorksheet = workbook.getWorksheet(2)

          // there should be a Products worksheet
          expect(worksheet).toBeDefined()
          // there should be only one worksheet with index 1
          expect(secondWorksheet).not.toBeDefined()
          expect(rows).toBeDefined()

          tempFile.removeCallback()
          done()
        })

        writer.writeToSingleXlsxFile(
          sampleStream,
          outputStream,
          logger,
          headers
        )
        done();
      })
    })

    test('handle empty rows', () => {
      return new Promise((done) => {
        const sampleStream = highland(sampleProducts)
        const tempFile = tmp.fileSync({ postfix: '.xlsx', keep: true })
        const output = tempFile.name
        const outputStream = fs.createWriteStream(output)
        const headers = ['id', 'key', 'productType']

        outputStream.on('error', done)
        outputStream.on('finish', async () => {
          const { rows } = await analyzeExcelFile(output)

          expect(rows).toHaveLength(4) // header + 3 products

          tempFile.removeCallback()
          done()
        })

        writer.writeToSingleXlsxFile(
          sampleStream,
          outputStream,
          logger,
          headers
        )
        done();
      })
    })

    test('log success info on xlsx completion', () => {
      return new Promise((done) => {
        const sampleStream = highland(sampleProducts)
        const headers = []
        const outputStream = streamTest.toText(() => { })
        outputStream.on('finish', () => {
          expect(logger.info).toHaveBeenCalledWith(
            expect.stringMatching(/written to XLSX file/)
          )
          done()
        })

        writer.writeToSingleXlsxFile(
          sampleStream,
          outputStream,
          logger,
          headers
        )
      })
    })
  })

  describe('::writeToZipFile', () => {
    test('write products to multiple files based on productTypes', () => {
      return new Promise((done) => {
        const sampleStream = highland(sampleProducts)
        const tempFile = tmp.fileSync({ postfix: '.zip', keep: true })
        const output = tempFile.name
        const outputStream = fs.createWriteStream(output)
        const entries = []

        // Extract data from zip file and test
        outputStream.on('finish', () => {
          fs.createReadStream(output)
            .pipe(unzipper.Parse())
            .on('entry', async (entry) => {
              const excelInfo = await analyzeExcelStream(entry)
              // push to entries array after we finish async operation
              // otherwise it would push both entries while analysing first entry
              // and the test would end prematurely
              entries.push(entry.path)

              // TODO (Node.js v18) - test content of excel files
              expect(excelInfo).toBeDefined()
              expect(excelInfo.rows).toBeDefined()

              // test if both productTypes were exported
              if (entries.length === 2) {
                expect(entries.sort()).toEqual([
                  'products/product-type-1.xlsx',
                  'products/product-type-2.xlsx',
                ])
                expect(entries).toHaveLength(2)
                tempFile.removeCallback()
                done()
              }
            })
        })

        writer.writeToZipFile(sampleStream, outputStream, logger)
        done()
      })
    })

    test('should handle exporting zero products', () => {
      return new Promise((done) => {
        const sampleStream = highland([])
        const tempFile = tmp.fileSync({ postfix: '.zip', keep: true })
        const output = tempFile.name
        const outputStream = fs.createWriteStream(output)
        let entries = 0

        // Extract data from zip file and test
        outputStream.on('finish', () => {
          fs.createReadStream(output)
            .pipe(unzipper.Parse())
            .on('entry', () => {
              entries += 1
              throw new Error('No entries should be archived')
            })
            .on('finish', () => {
              // there should be no products in a result zip file
              expect(entries).toEqual(0)
              tempFile.removeCallback()
              done()
            })
        })

        writer.writeToZipFile(sampleStream, outputStream, logger)
      })
    })

    test('log success info on zip completion', () => {
      return new Promise((done) => {
        const sampleStream = highland(sampleProducts)

        const tempFile = tmp.fileSync({ postfix: '.zip' })
        const output = tempFile.name
        const outputStream = fs.createWriteStream(output)

        outputStream.on('finish', () => {
          expect(logger.info).toHaveBeenCalledWith(
            expect.stringMatching(/written to ZIP file/)
          )
          tempFile.removeCallback()
          done()
        })

        writer.writeToZipFile(sampleStream, outputStream, logger)
      })
    })

    test('throw an error when streams fail', () => {
      return new Promise((done) => {
        const tempFile = tmp.fileSync({ postfix: '.zip' })
        const tmpDir = tmp.dirSync({ unsafeCleanup: true })
        const outputStream = fs.createWriteStream(tempFile.name)
        const failedStream = fs.createWriteStream(tempFile.name)
        const excelExports = [
          {
            excel: {
              stream: failedStream,
              finish: () => failedStream.emit('error', new Error('test error')),
            },
          },
        ]

        outputStream.on('error', (err) => {
          expect(err).toBeDefined()
          expect(err.message).toEqual('test error')

          tempFile.removeCallback()
          tmpDir.removeCallback()
          done()
        })

        writer.finishWorksheetsAndArchive(
          excelExports,
          tmpDir.name,
          outputStream,
          logger
        )
      })
    })
  })
})
