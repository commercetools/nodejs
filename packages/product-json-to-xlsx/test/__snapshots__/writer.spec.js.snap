// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Writer ::writeToSingleXlsxFile handle empty rows 1`] = `
Array [
  Array [
    "id",
    "key",
    "productType",
  ],
  Array [
    "12345ab-id",
    "product-key",
    "product-type-1",
  ],
  Array [
    "75345ab-id",
    "product-key-2",
    "product-type-2",
  ],
  Array [
    "another5ab-id",
    "product-key",
    "product-type-1",
  ],
]
`;

exports[`Writer ::writeToSingleXlsxFile write products to a single file with specified headers 1`] = `
Array [
  Array [
    "id",
    "key",
    "productType",
    "variantId",
    "sku",
    "images",
    "anotherAddedAttr",
    "article",
    "designer",
    "color",
  ],
  Array [
    "12345ab-id",
    "product-key",
    "product-type-1",
    1,
    "A0E200000001YKI",
    "image-url1",
    undefined,
    "sample 089 WHT",
    undefined,
    "white",
  ],
  Array [
    undefined,
    undefined,
    undefined,
    6,
    "A0E200001YKI123",
    "https://image1.de;myImage.com",
    undefined,
    "sample 089 WHT",
    undefined,
    "white",
  ],
  Array [
    "75345ab-id",
    "product-key-2",
    "product-type-2",
    1,
    "A0E200000001YKI",
    "https://example.com/foobar/commer.jpg|3|4",
    undefined,
    "sample 089 WHT",
    "michaelkors",
    "white",
  ],
  Array [
    undefined,
    undefined,
    undefined,
    3,
    "A0E200001YKI123",
    "https://example.com/foobar/commer234.jpg|3|3",
    undefined,
    undefined,
    "michaelkors",
  ],
  Array [
    "another5ab-id",
    "product-key",
    "product-type-1",
    1,
    "A0E200001YKI",
    undefined,
    undefined,
    "sample 777 WHT",
    undefined,
    "grune",
  ],
  Array [
    undefined,
    undefined,
    undefined,
    21,
    "PPP0001YKI123",
    "https://eg993",
    undefined,
    undefined,
    undefined,
    "blau",
  ],
]
`;

exports[`Writer ::writeToZipFile write products to multiple files based on productTypes: xlsx1 1`] = `
Array [
  Array [
    "id",
    "key",
    "productType",
    "variantId",
    "sku",
    "images",
    "addedAttr",
    "anotherAddedAttr",
    "article",
    "color",
    "colorFreeDefinition.en",
  ],
  Array [
    "12345ab-id",
    "product-key",
    "product-type-1",
    1,
    "A0E200000001YKI",
    "image-url1",
    undefined,
    undefined,
    "sample 089 WHT",
    "white",
    "black-white",
  ],
  Array [
    undefined,
    undefined,
    undefined,
    6,
    "A0E200001YKI123",
    "https://image1.de;myImage.com",
    undefined,
    undefined,
    "sample 089 WHT",
    "white",
    "black-white",
  ],
  Array [
    "another5ab-id",
    "product-key",
    "product-type-1",
    1,
    "A0E200001YKI",
    undefined,
    undefined,
    undefined,
    "sample 777 WHT",
    "grune",
    "grey-white",
  ],
  Array [
    undefined,
    undefined,
    undefined,
    21,
    "PPP0001YKI123",
    "https://eg993",
    undefined,
    undefined,
    undefined,
    "blau",
    "schwarz-weiß",
  ],
]
`;

exports[`Writer ::writeToZipFile write products to multiple files based on productTypes: xlsx2 1`] = `
Array [
  Array [
    "id",
    "key",
    "productType",
    "variantId",
    "sku",
    "images",
    "addedAttr",
    "anotherAddedAttr",
    "article",
    "color",
    "colorFreeDefinition.en",
    "colorFreeDefinition.de",
    "designer",
  ],
  Array [
    "75345ab-id",
    "product-key-2",
    "product-type-2",
    1,
    "A0E200000001YKI",
    "https://example.com/foobar/commer.jpg|3|4",
    undefined,
    undefined,
    "sample 089 WHT",
    "white",
    "black-white",
    "schwarz-weiß",
    "michaelkors",
  ],
  Array [
    undefined,
    undefined,
    undefined,
    3,
    "A0E200001YKI123",
    "https://example.com/foobar/commer234.jpg|3|3",
    undefined,
    undefined,
    undefined,
    undefined,
    undefined,
    "schwarz-weiß",
    "michaelkors",
  ],
]
`;
